sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -470,14 +470,36 @@ case class AlterTableAddPartitionCommand(
       CatalogTablePartition(normalizedSpec, table.storage.copy(
         locationUri = location.map(CatalogUtils.stringToURI)))
     }
-    catalog.createPartitions(table.identifier, parts, ignoreIfExists = ifNotExists)
+
+    // Hive metastore may not have enough memory to handle millions of partitions in single RPC.
+    // Also the request to metastore times out when adding lot of partitions in one shot.
+    // we should split them into smaller batches
+    val batchSize = sparkSession.sessionState.conf.partitionBatchSize
+    parts.toIterator.grouped(batchSize).foreach { batch =>
+      catalog.createPartitions(table.identifier, batch, ignoreIfExists = ifNotExists)
+    }
 
     if (table.stats.nonEmpty) {
       if (sparkSession.sessionState.conf.autoSizeUpdateEnabled) {
-        val addedSize = parts.map { part =>
-          CommandUtils.calculateLocationSize(sparkSession.sessionState, table.identifier,
-            part.storage.locationUri)
-        }.sum
+        val threshold = sparkSession.sparkContext.conf.get(RDD_PARALLEL_LISTING_THRESHOLD)
+        val evalPool = ThreadUtils.newForkJoinPool("AlterTableAddPartitionsCommand", 8)
```

Comment:
I'd rewrite this to not create a pool unless it's going to be used.

Reply:
Also can this use ThreadUtils.parmap instead? might be cleaner

Reply:
done. Using ThreadUtils.parmap now.